### DISCUSSION TITLE
Installing micropython-esp32-ulp with mip on latest firmware esp32

### DISCUSSION THREAD

**Discussion by mjaspers2mtu** [micropython/micropython#12184] [ESP32] [answered]
Hello, I am trying to use the ULP on my esp32.

For this I wish to install this: https://github.com/micropython/micropython-esp32-ulp/tree/master

However the simple instruction of 
```py
import upip
upip.install('micropython-esp32-ulp')
```
does not work. I Found that this is because upip has been replaced by mip.
However, with mip, it is unable to locate this package.
I have tried:
```py
import mip
mip.install('micropython-esp32-ulp')
mip.install('esp32-ulp')
```
```
Installing esp32-ulp (latest) from https://micropython.org/pi/v2 to /lib
Package not found: https://micropython.org/pi/v2/package/6/esp32-ulp/latest.json
Package may be partially installed
```

Can someone please explain how I can get esp32-ulp installed on '3.4.0; MicroPython v1.20.0-327-gd14ddcbdb on 2023-07-27'

**Comment by karfas** [comment]
I have not tried this, but according to the [documentation](https://docs.micropython.org/en/latest/reference/packages.html), it is possible to install mip packages directly from github using the syntax
```py
mip.install("github:org/repo/path/foo.py")  # Uses default branch
mip.install("github:org/repo/path/foo.py", version="branch-or-tag")  # Optionally specify the branch or tag
```

**Comment by peterhinch** [comment]
It's not obvious how to install this. For a package to be installable with `mip` it needs a `package.json` file and I can't see one in the GitHub repo. 

Your best bet is to go to PyPi and download the files from [here](https://pypi.org/project/micropython-esp32-ulp/#files) then copy them to the ESP32. I'm guessing that what's on PyPi matches what would have been installed with upip.

Good luck! You might like to report back if you have success as I'm sure others would like to use the ULP. Another thought is to raise an issue: the maintainers of the library might like to make it mip-installable. It's not a big job.

**Reply by mjaspers2mtu** [reply, marked as answer]
Thanks for the reply and looking into an alternative way to get this installed. I followed your tip and indeed created an issue: 
https://github.com/micropython/micropython-esp32-ulp/issues/93

**Comment by wnienhaus** [comment]
ICYMI: MIP support was added in this PR: https://github.com/micropython/micropython-esp32-ulp/pull/94
You can now install `micropython-esp32-ulp` with:
```py
import mip
mip.install('github:micropython/micropython-esp32-ulp')
```
